### DISCUSSION TITLE
Include multiple portman config json

### DISCUSSION THREAD

**Discussion by jdevosMagento** [apideck-libraries/portman#274] [General]
Hello,

First of all, thank you for this tool which is really interesting! 

I wanted to know if it was possible to make file inclusions in a root configuration file.

I have a big project in terms of api and I'm afraid that having everything in a single file will become unreadable.

With openapi yaml, I was able to cut out the different paths via $refs which allows you to cut out the interfaces well and find your way around easily. 

Is there a similar solution with portman?

If not, what advice can you give to get organized on a project with multiple APIs? 

Thank you!

**Comment by thim81** [comment]
Hi @jdevosMagento 

I think you are referring to splitting up the Portman configuration (or the OpenAPI file)?

For the Portman conversion configuration, currently we expect it all grouped in 1 file.

You can include a file reference for the [prerequestscripts](https://github.com/apideck-libraries/portman#portman---operationprerequestscripts-properties) using a "file:" syntax, to make the configuration a bit more compact.

### Possible workaround
You can define the Portman configuration in JSON or YAML, what you could do is add an additional step before you run the portman conversion by leveraging a dereference package (ex: [json-schema-ref-parser]( https://github.com/APIDevTools/json-schema-ref-parser)).

1. Dereference your your Portman config (`portman-config.yaml`) & save the result as a new file (`portman-config.bundled.yaml`)
2. Pass the bundled file to Portman `portman -l your-openapi-file.yaml -c portman-config.bundled.yaml`

**Reply by jdevosMagento** [reply]
Thank you for your answer, i will look at it

**Comment by thim81** [comment]
Curious how large you Portman config would be?
And if more members from the community struggle with large Portman configurations.

**Reply by jdevosMagento** [reply]
It's a big Magento e-commerce platform, with native APIs as well as specific ones to connect to several external bricks, I can't say how many exactly but there are a lot.

**Comment by thim81** [comment]
Tagging @savage-alex & @AliKhalili & @nicklloyd to capture their experience here.

**Reply by AliKhalili** [reply]
I have a two different Portman configurations, first one is general configuration which I use for generating a set of common tests for all the APIs regardless of their internal business, and the second one is specific configuration which I create and use for each OAS3 separately. So, it would be a great idea to have something like OAS bundle for Portman configurations. But it is worth mentioning that more feature more thing to learn, more liability. I prefer a simple tools with a few well-defined features.

**Reply by thim81** [reply]
@AliKhalili What if it is supported by default, so that it does not require any thoughts or implementation from the users? 
Like OpenAPI works. You can use $ref (for easier maintenance and re-use) but you can also just keep the single file as we have right now and create a straightforward config without $ref. Portman could dereference any Portman config before the actual conversion by default.

**Reply by AliKhalili** [reply]
That is quite interesting if I can use $ref in my Portman config. it can solve growth issue of config file and at the same time we don't need to care about the maintaining different config files. Portman does that for us.

**Reply by jdevosMagento** [reply]
yes its interesting

**Reply by thim81** [reply]
Ok, given the feedback, this is a valuable feature addition for Portman.
A [work item ](#277) is created. Whenever we find time, we will be working on the PR for this feature.

**Reply by savage-alex** [reply]
sorry for delay in response. I have been off on sick leave. This approach sounds great!

**Comment by thim81** [comment]
@AliKhalili @jdevosMagento 
The PR is in the making: https://github.com/apideck-libraries/portman/tree/277-%24ref/examples/config-references

If you have some time, have a look and see if the proposal would work?
The PR lacks some improved error handling, which will be added soon.

**Reply by AliKhalili** [reply]
@thim81 Fantastic job, well-done. It's my pleasure. I'll do that for the upcoming weekend.

**Comment by thim81** [comment]
hi @AliKhalili, @savage-alex & @jdevosMagento We just released version 1.15.0 of Portman, where this feature is included.

**Comment by thim81** [comment]
If in the coming weeks, anybody will give the '$ref' usage an attempt, if possible leave some feedback (good or bad).

**Reply by AliKhalili** [reply]
@thim81 I used this new feature, it've worked as we expected, thank you again.

**Reply by jdevosMagento** [reply]
Hello @thim81, i tested last week and it seems to work pretty well! I'm continuing the integration this week, I'll let you know.

**Comment by thim81** [comment]
@AliKhalili @jdevosMagento Are you still using the `$ref` option for simplifying the Portman configurations?

Since Portman config can become quite big, especially when using Variations & Fuzzing, we are wondering how we could document this capability better so that users would discover the `$ref` option to streamline their Portman configs?

Any tips or suggestion are welcome.

**Reply by AliKhalili** [reply]
Sorry @thim81 for delay. During the last months I have no chance to work in Portman. At that time, it worked for me very well. I am going to ask my collage to share their thoughts here.
